### PR TITLE
Added Alias method for sampling multinomial

### PR DESCRIPTION
**Pull request by superbobry**
Unlike the previous cumsum implementation, [the Alias method](http://www.keithschwarz.com/darts-dice-coins) generates a value in constant time.

I also wanted to discuss the naming of `multinomial`. The sampling procedure looks like it samples from a discrete (aka [Categorical](https://en.wikipedia.org/wiki/Categorical_distribution)) distribution.

### CODE REVIEW

File: src/lib/stats/sampling.ml [hunk 1]
```diff
@@ -61,27 +62,76 @@ let normal ?seed ~mean ~std () =
...
+        else if weights.(i) < average then
+          partition weights (succ i) (i::small, large)
+        else
+          partition weights (succ i) (small, i::large)
+      in
+
+      let (small, large) = partition weights 0 ([], [])
+      and probs = Array.make n 0.0
+      and alias = Array.make n 0 in begin
+        iter weights probs alias (small, large);
```
Comment (rleonid) [inline src/lib/stats/sampling.ml:112]: I used to also use `and` when declaring multiple items but from feedback from other developers I've reserved it for mutually recursive cases. This way I know that there is actually something special and complicated about an `and` binding.

File: src/lib/stats/sampling.ml [hunk 2]
```diff
@@ -61,27 +62,76 @@ let normal ?seed ~mean ~std () =
...
+        { probs; alias }
+      end
+
+    let sample { probs; alias } r =
+      let i = Random.State.int r (Array.length alias) in
```
Comment (rleonid) [inline src/lib/stats/sampling.ml]: I tend to declare and pass around array lengths.

File: src/lib/stats/sampling.ml [hunk 3]
```diff
@@ -61,27 +62,76 @@ let normal ?seed ~mean ~std () =
...
+      if (Random.State.float r 1.0 < probs.(i)) then i else alias.(i)
+  end
+
 let multinomial ?seed weights =
   if not (Array.all (within (Open 0.0, Closed 1.0)) weights) then
-    raise (Invalid_argument "weights")
+    invalid_arg ~f:"multinomial" "weights must be within (0, 1]"
   else
```
Comment (rleonid) [inline src/lib/stats/sampling.ml:124]: Thanks for converting these to use `invalid_arg` !

File: src/lib/stats/sampling.ml [hunk 4]
```diff
@@ -61,27 +62,76 @@ let normal ?seed ~mean ~std () =
   let rsn = normal_std ?seed () in
   (fun () -> std *. (rsn ()) +. mean)
 
+(** Alias method for sampling from a categorical distribution.
+
+    The method allows to generate a sample in costant time after a
+    linear time preprocessing of the probabilities array. *)
+module Alias =
```
Comment (rleonid) [inline src/lib/stats/sampling.ml:70]: I would also add a link to the source of the algorithm in this comment. Either Vose's paper or the Kevin Schwartz website would be great.

**Comment by rleonid** [comment]
Thank you for your contribution and introducing me to such an elegant algorithm! 

Regarding the naming, you're correct. Feel free to file an issue or fix the incorrect usage in this PR. I'll merge this PR once a bit more testing on the 4.03 logic is complete (mostly just built systems checking).

I've also added a random test in this [commit](https://github.com/hammerlab/oml/commit/2e13fb1c1485affdd1e54a1213cf624035e633c5).

Although in general I am pretty loose about style guidelines, I've made a couple of comments that you can ignore for now and maybe take into account in the future.

**Comment by superbobry** [comment]
Thanks for the review! I've addressed all of the comments and additionally renamed Multinomial to Categorical.

**Pull request merged**